src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+            Dictionary<string, Object> subDim = GetSubDim(dimension, dimensionInfo.DetailDims);
+
+            foreach (Point leave in pointTree.Leaves)
+            {
+                if (ContainsAll(leave.Dimension, subDim))
+                {
+                    count++;
+
+                    p.Value = +leave.Value;
+                    p.ExpectedValue = +leave.ExpectedValue;
+                    p.Delta = +leave.Delta;
+                }
+
+            }
+            if (aggType.Equals(AggregateType.Avg))
+            {
+                p.Value = p.Value / count;
+                p.ExpectedValue = p.ExpectedValue / count;
+                p.Delta = p.Delta / count;
+            }
+
+            if (count > 0)
+            {
+                return p;
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private static string GetDicCode(Dictionary<string, Object> dic)
+        {
+            return string.Join(";", dic.Select(x => x.Key + "=" + (string)x.Value).ToArray());
+        }
+
+        private void BuildTree(PointTree tree, List<string> aggDims, Point point, Object aggSymbol)
+        {
+            int aggNum = 0;
+            string nextDim = null;
+
+            foreach (string dim in aggDims)
+            {
+                if (IsAggregationDimension(point.Dimension[dim], aggSymbol))
+                {
+                    aggNum++;
+                }
+                else
+                {
+                    nextDim = dim;
+                }
+            }
+
+            if (aggNum == aggDims.Count)
+            {
+                tree.ParentNode = point;
+            }
+            else if (aggNum == aggDims.Count - 1)
+            {
+                if (!tree.ChildrenNodes.ContainsKey(nextDim))
+                {
+                    tree.ChildrenNodes.Add(nextDim, new List<Point>());
+                }
+                tree.ChildrenNodes[nextDim].Add(point);
+            }
+
+            if (aggNum == 0)
+            {
+                tree.Leaves.Add(point);
+            }
+        }
+
+        private BestDimension FindBestDimension(SortedDictionary<BestDimension, double> valueMap, Dictionary<BestDimension, double> valueRatioMap, double meanGain, bool isLeavesLevel = true)
+        {
+            BestDimension best = null;
+            foreach (KeyValuePair<BestDimension, double> dimension in valueMap)
+            {
+                if (dimension.Key.AnomalyDis.Count == 1 || (isLeavesLevel ? dimension.Value >= meanGain : dimension.Value <= meanGain))
+                {
+                    if (dimension.Key.AnomalyDis.Count > 1)
+                    {
+                        if (best == null || (!Double.IsNaN(valueRatioMap[best]) && (best.AnomalyDis.Count != 1 && (isLeavesLevel ? valueRatioMap[best].CompareTo(dimension.Value) <= 0 : valueRatioMap[best].CompareTo(dimension.Value) >= 0))))
+                        {
```

Comment:
suggestion: For readability, can you please avoid the long line and break up the conditions on separate lines? Same comment for line 487.

Reply:
> suggestion: For readability, can you please avoid the long line and break up the conditions on separate lines? Same comment for line 487.

made some updates